Adding functionality to check for 'no-scan' countries

st0rmbl3ss3d opened the pull request:
## 🗣 Description ##

This is a change designed to check the GeoIP db for IPs in countries that are currently not supposed to be scanned per policy. 

## 💭 Motivation and context ##

This change is required because we recently officially updated our scanning policy across VM to exclude scans to IPs located in certain countries. This change solves the problem by adding an automatic check to look up the country for each CIDR that is either imported or added to avoid importing IPs that should not be scanned.

## 🧪 Testing ##

I tested this change by forking my own repo and then generating a local docker instance, with the actual importing steps that would change the DB commented out. I then made a dummy json file and attempted to import it with known IPs that belong to those countries. I also tested the 'add' functionality in the same way, commenting out the actual addition, but attempting to add known IPs that belong to said countries. The tests were successful as I tried them.

st0rmbl3ss3d committed:
Update cyhy-import

Added functionality to check the geoip database for countries that shouldn't be scanned. Did this by adding new function 'has_restricted' that returns false if it finds an address in a 'no-scan' country then checking for that, similar to the intersections check.

st0rmbl3ss3d committed:
Update cyhy-ip

Added functionality to check added ips in the add function for ips in 'no-scan' countries. Added new function 'has_restricted_ips' to check for ips in 'no-scan' countries.

jsf9k commented on bin/cyhy-ip:
```diff
@@ -163,6 +179,8 @@ def add(db, owner, cidrs):
         cidrs
     )  # intersections from database
+    if has_restricted_ips(cidrs):
+        print("Cannot continue!\nSome addresses are in invalid countries.") # Current countries that are not to be scanned.
```
I think you want a `sys.exit(-1)` or something similar here.  Otherwise processing won't stop.

st0rmbl3ss3d replied:
Ah, good catch. I thought I added that. Will do.

jsf9k commented on bin/cyhy-ip:
```diff
@@ -110,7 +111,22 @@ def get_special_intersections(cidrs):
     return results
 
-
+def has_restricted_ips(nets):
+    reader = geoip2.database.Reader('/usr/local/share/GeoIP/GeoIP2-City.mmdb')
+    bad_countries = ['China','Russia','Iran','North Korea']
+    has_bad_address = False
+
+    for cidr in nets.iter_cidrs():
+        geo_data = reader.city(str(cidr[0]))
+        country_name = geo_data.country.name
+        if country_name in bad_countries:
+            print('{} contains addresses in country that is unavailable due to current scanning policy: {}'.format(cidr,country_name))
+            has_bad_address = True
+    if has_bad_address:
+        return True
+    else:
+        return False
+  
```
I don't like the same code being repeated in two places.  Is there any way to put this code in a common place?

st0rmbl3ss3d replied:
Yeah, I didn't like copying the code over twice but I am trying to think of how to do it so both files can access it.

jsf9k replied:
We'll see if @cisagov/team-ois (the other devs) have any ideas as well.  I'm not very familiar with this part of the CyHy code, so I don't know the best place to move it to.

mcdonnnj replied:
Is there any reasons not to add a single CIDR argument method to `GeoLocDB` in https://github.com/cisagov/cyhy-core/blob/develop/cyhy/core/geoloc.py ? That class is already in use by both tools, and adding it there would make it available for any future needs as part of the library.

dav3r replied:
I am on board with @mcdonnnj's suggestion of adding a method to [`GeoLocDB`](https://github.com/cisagov/cyhy-core/blob/develop/cyhy/core/geoloc.py) to handle this lookup- something like`check_restricted_cidr()`. 👍

st0rmbl3ss3d replied:
Yes, that seems like a good solution

st0rmbl3ss3d replied:
Added a commit to my branch with the requested changes made. Does this look good?

jsf9k reviewed (changes requested):
I left two comments.  I'd like to see the `has_restricted_ips()` function defined in only one place, if possible.  A bigger bug is that `cyhy-ip` prints a message but does not actually stop executing when the `has_restricted_ips()` function returns `True`.

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -79,6 +80,21 @@ def has_intersections(db, nets, filename, owner):
 
+def has_restricted_ips(nets):
+    reader = geoip2.database.Reader('/usr/local/share/GeoIP/GeoIP2-City.mmdb')
+    bad_countries = ['China','Russia','Iran','North Korea']
+    has_bad_address = False
```
I don't like the use of "good" and "bad" when involving real world entities as they're words that represent value judgements. I suggested restricted as a replacement as it aligns with the method name you chose.

```suggestion
    restricted_countries = ['China','Russia','Iran','North Korea']
    has_restricted_address = False
```

jsf9k replied:
Good suggestion!

st0rmbl3ss3d replied:
Yes, I like that. I am bad at coming up with variable names haha

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -79,6 +80,21 @@ def has_intersections(db, nets, filename, owner):
+        country_name = geo_data.country.name
+        if country_name in bad_countries:
+            print('{} contains addresses in country that is unavailable due to current scanning policy: {}'.format(cidr,country_name))
+            has_bad_address = True
+    if has_bad_address:
+        return True
+    else:
+        return False
```
Updating variable name for consistency and since it's a boolean you can just return it instead of using an if/else.

```suggestion
            has_restricted_address = True

    return has_restricted_address
```

mcdonnnj reviewed (changes requested):
Implementation makes sense. Unless there's disagreement about adding a method to `GeoLocDB`, I would like to see that change to DRY the code and provide that functionality for other purposes. I did have some changes below I would like to see.

st0rmbl3ss3d committed:
Update bin/cyhy-import

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-import

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
made requested changes by adding the checkign functionalioty to GeoIP and updating language. Fixed bug wher eprocess did not exit on finding a restricted IP

st0rmbl3ss3d committed:
Merge branch 'develop' of https://github.com/st0rmbl3ss3d/cyhy-core into develop

st0rmbl3ss3d committed:
removed unnecessary import

st0rmbl3ss3d committed:
Accidentally uncommented a line I wasn't supposed to.

jsf9k commented on .bash_history:
```diff
@@ -0,0 +1,3 @@
+cd /usr/local/share/GeoIP/
+ls
+exit
```
This file needs to be removed.  I assume it was committed in error.

jsf9k commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +39,15 @@ def lookup(self, ip):
             return (None, None)
+
+    def check_restricted_cidr(self, cidr):
+        restricted_countries = ["China","Iran","North Korea","Russia"]
```
The `restricted_countries` variable should be defined at the top of the file as `RESTRICTED_COUNTRIES`, similar to what is done with `GEODB_FILES`.

jsf9k commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +39,15 @@ def lookup(self, ip):
+                print("Warning! %s traced to restricted country: %s" % (cidr, response.country.name))
+                has_restricted = True
+        except AddressNotFoundError:
+            print('Address not found.')
```
This error message is not very descriptive.  How about
```suggestion
            print("CIDR %s not found in geolocation database", cidr)
```
instead?

st0rmbl3ss3d replied:
Yes, more descriptive would be better. Will note.

jsf9k commented on bin/cyhy-ip:
```diff
@@ -163,6 +162,13 @@ def add(db, owner, cidrs):
     )  # intersections from database
+    geo_loc_db = GeoLocDB()
+    for cidr in cidrs.iter_cidrs():
+        has_restricted = geo_loc_db.check_restricted_cidr(cidr)
+        if has_restricted:
```
```suggestion
        if geo_loc_db.check_restricted_cidr(cidr):
```

jsf9k commented on bin/cyhy-import:
```diff
@@ -79,6 +80,16 @@ def has_intersections(db, nets, filename, owner):
 
+def has_restricted_ips(nets):
+    geo_loc_db = GeoLocDB()
+    has_restricted = False
+    for cidr in nets.iter_cidrs():
+        if geo_loc_db.check_restricted_cidr(cidr):
+            has_restricted = True
+    if has_restricted:
+        return True
+    else:
+        return False
```
This logic can be simplified considerably:
```suggestion
    for cidr in nets.iter_cidrs():
        if geo_loc_db.check_restricted_cidr(cidr):
            return True
    return False
```

st0rmbl3ss3d replied:
The reason I had it like this was so it had the chance to iterate through all the cidrs before returning, as the check function in GeoDB prints out the matching IPs, in case there were multiples so the user would not have to keep going back. Is there an easier way to do this?

mcdonnnj replied:
This would simplify while still allowing it to iterate through all of the CIDRs:

```suggestion
    has_restricted = False
    for cidr in nets.iter_cidrs():
        if geo_loc_db.check_restricted_cidr(cidr):
            has_restricted = True

    return has_restricted
```

jsf9k replied:
See [here](https://github.com/cisagov/cyhy-core/pull/60#discussion_r595360456) for my response.

jsf9k reviewed (changes requested):
Great work!  There are a few places that can still be improved a little.

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

st0rmbl3ss3d committed:
Moved restricted_countries to the global variables, changed the variable name to all caps to be in-line with other globals. Also updated the reference to the variable to the new variable name.

jsf9k commented on bin/cyhy-import:
```diff
@@ -79,6 +80,16 @@ def has_intersections(db, nets, filename, owner):
+    if has_restricted:
+        return True
+    else:
+        return False
```
OK, since you want the side-effects we can only partially simplify it:
```suggestion
    has_restricted = False
    for cidr in nets.iter_cidrs():
        if geo_loc_db.check_restricted_cidr(cidr):
            # We can't just exit here since check_restricted_cidr()
            # prints output for each restricted CIDR block and we
            # want the user to see all restricted CIDR blocks
            # without having to run multiple times.
            has_restricted = True
    return has_restricted
```

jsf9k commented on bin/cyhy-import:
```diff
@@ -93,6 +104,8 @@ def import_request(db, request, source, force=False, init_stage=None):
         request["init_stage"] = init_stage
     if has_intersections(db, nets, source, owner):
         return False
+    if has_restricted_ips(nets):
+        return False
```
I assume you don't use an `or` here because that might short circuit and you want the user to see any output that is printed by each function.  You should put a comment here explaining that (and anywhere else you do something similar).  Otherwise I will come through on the next PR and suggest combining the two function calls via `or` because I have forgotten why it's written the way it is.

st0rmbl3ss3d replied:
Correct. I will add in that comment

jsf9k reviewed (changes requested):
A few more suggestions.  Fear not, we are converging.

st0rmbl3ss3d committed:
Update bin/cyhy-import

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

st0rmbl3ss3d committed:
Added comment to indicate necessity of having separate checks instead of combining with an 'or' statement.

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -146,6 +163,9 @@ def main():
 
     if not success:
         sys.exit(-1)
+    else:
+        print("Success.")
```
```suggestion

    print("Success.")
```

st0rmbl3ss3d replied:
Oops, this statement can actually be removed. I only added it in when I commented out the actual importing and adding of IPs when I was doing testing so I can know whether the import would have been successful or not. If you all would like to keep the success message then that's cool, but it wasn't there before and isn't needed for this change

mcdonnnj replied:
I'm fine with removing if it's not something your team needs.

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -93,6 +105,11 @@ def import_request(db, request, source, force=False, init_stage=None):
     if has_intersections(db, nets, source, owner):
         return False
+    if has_restricted_ips(nets):
+        #This is not combined with the previous check so the output
+        #from each function can be seen so the user can get all messages
```
```suggestion
        # This is not combined with the previous check so the output
        # from each function can be seen so the user can get all messages
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +162,12 @@ def add(db, owner, cidrs):
     )  # intersections from database
+    geo_loc_db = GeoLocDB()
+    for cidr in cidrs.iter_cidrs():
+        if geo_loc_db.check_restricted_cidr(cidr):
+            print("Cannot continue!\nSome addresses are in restricted countries.")
+            sys.exit(-1)
```
This won't check every CIDR, Is this desired behavior?

st0rmbl3ss3d replied:
This is within a 'for cidr in cidrs.iter_cidrs(): statement. Does that not loop through the cidrs as expected? If not, we can change the logic.

mcdonnnj replied:
When you call `sys.exit(-1)` you will "bail out" at the first restricted CIDR.

st0rmbl3ss3d replied:
Ahh yeah.  I will see what I can do to resolve

jsf9k replied:
@mcdonnnj, I think this is resolved now.  Do you agree?

mcdonnnj replied:
![image](https://user-images.githubusercontent.com/50747025/111652363-68c20500-87dd-11eb-8b2f-dc71418b3347.png)

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +40,14 @@ def lookup(self, ip):
+        try:
+            response = self.__reader.city(str(cidr[0]))
+            if response.country.name in RESTRICTED_COUNTRIES:
+                print("Warning! %s traced to restricted country: %s" % (cidr, response.country.name))
+                has_restricted = True
+        except AddressNotFoundError:
+            print("CIDR %s not found in geolocation database", cidr)
```
I think I would prefer these to print to `sys.stderr` as most of the `print` statements in this library do, but what does everyone else think @cisagov/team-ois ?

jsf9k replied:
Agreed.  Great catch!

jsf9k replied:
I think @mcdonnnj's intent was to print _all_ these error messages to `stderr`, not just the one example he pointed out.

mcdonnnj replied:
I did, and I would prefer them to follow the format seen elsewhere in the code base such as https://github.com/cisagov/cyhy-core/blob/79d25173b91c6cd71bd106733f21020c81282e5a/cyhy/db/database.py#L130

jsf9k replied:
@mcdonnnj, I'm waiting for this last request to be resolved before re-reviewing.  I'll likely approve if nothing else comes up; @st0rmbl3ss3d has done some useful and quality work here!

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +40,14 @@ def lookup(self, ip):
+                print("Warning! %s traced to restricted country: %s" % (cidr, response.country.name))
+                has_restricted = True
+        except AddressNotFoundError:
+            print("CIDR %s not found in geolocation database", cidr)
```
```suggestion
            print("CIDR %s not found in geolocation database" % cidr)
```

mcdonnnj reviewed (changes requested):
I had some more changes and a couple questions. Thanks for continuing to work on getting this into an approved state!

st0rmbl3ss3d committed:
Update bin/cyhy-import

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Made some of the suggested changes. Please look through to see if they resolv current concerns.

dav3r reviewed (approved):
This all looks good to me now. 👍  Nice work @st0rmbl3ss3d!

jsf9k commented:
@st0rmbl3ss3d - in the future please try to create `git commit` messages with more descriptive content, even when committing suggested changes.  The default message of "Update bin/cyhy-import", for example, does not tell the reader anything about what changes are being made.

st0rmbl3ss3d committed:
Updated all print statements to instead print to stderr in the format recommended by @mcdonnj. Specifically, lines 49 and 52 in geoloc.py, line 171 in cyhy-ip were updated. Also added a warning message to the overall failure in cyhy-import on lines 93-34 as there wasn't one before and it relied only on the error messages from geoloc.py. Also updated the formatting and wording of the error messages for consistency and corrected a minor spelling mistake (counties > countries)

jsf9k reviewed (approved):
Strong work @st0rmbl3ss3d!

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -79,6 +80,19 @@ def has_intersections(db, nets, filename, owner):
+            # without having to run multiple times.
+            has_restricted = True
+    if has_restricted:
+        print >> sys.stderr,"Cannot import.\nSome addresses traced to restricted countries."
```
```suggestion
        print "Cannot import.\nSome addresses traced to restricted countries."
```

st0rmbl3ss3d replied:
Ok then, I must have been confused earlier when you said they should all be going to stderr, perhaps it was just with geoloc.py. my bad. I will commit these suggestions to get them in line.

mcdonnnj replied:
Yes, sorry. It was specifically for `geoloc.py` to conform with other code in the `cyhy-core` library.

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +162,15 @@ def add(db, owner, cidrs):
+        if geo_loc_db.check_restricted_cidr(cidr):
+            has_restricted_ips = True
+    if has_restricted_ips:
+        print >> sys.stderr,"Cannot continue!\nSome addresses traced to restricted countries."
```
```suggestion
        print "Cannot continue!\nSome addresses traced to restricted countries."
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -172,7 +180,7 @@ def add(db, owner, cidrs):
     # get the init_stage, if it doesn't have one assume NETSCAN1
     stage = STAGE[request.get("init_stage", STAGE.NETSCAN1)]
-    geo_loc_db = GeoLocDB()
+
```
```suggestion
```

mcdonnnj reviewed (changes requested):
We're getting there! I am requesting removal of the redirect to `sys.stderr` in the scripts because it is inconsistent with the other print statements in those scripts. Appreciate your work dealing with our feedback!

st0rmbl3ss3d committed:
Update bin/cyhy-import

changing the redirect to sys.stderr in this print statement to a simple print statement to be more in line with how the rest of the print statements are handled.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

Likewise changing this direct to stderr to a print statement to align with standard usage throughout the script

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

Looks like some extra whitespace was added. Committing this change of whitespace removal for formatting purposes

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +162,15 @@ def add(db, owner, cidrs):
+    has_restricted_ips = False
+    for cidr in cidrs.iter_cidrs():
+        if geo_loc_db.check_restricted_cidr(cidr):
+            has_restricted_ips = True
```
Explanatory comments should be here as well since the reason is the same.

```suggestion
            # We can't just exit here since check_restricted_cidr()
            # prints output for each restricted CIDR block and we
            # want the user to see all restricted CIDR blocks
            # without having to run multiple times.
            has_restricted_ips = True
```

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +40,14 @@ def lookup(self, ip):
+    def check_restricted_cidr(self, cidr):
+        has_restricted = False
+        try:
+            response = self.__reader.city(str(cidr[0]))
```
Looking at this again, this is only checking the first IP of an entire CIDR. I think that this method should take a single IP like the `lookup()` method, and the scripts should iterate over each IP in a CIDR to check.

jsf9k replied:
Great catch @mcdonnnj!

st0rmbl3ss3d replied:
Ah, you are right. I was thinking it would be enough that way but it should loop through them all to be sure. I will re-work the logic. My only thing is that I don't want to print out each and every IP that has an issue, as sometimes there will be /16, etc. So I will have it check all the IPs but only print the cidr if it is problematic. Less granularity of provided information, but it wont dump 65k IPs into the terminal. I will try and get that committed today for review.,

mcdonnnj reviewed (changes requested):
Sorry, just a couple more things.

st0rmbl3ss3d committed:
Update bin/cyhy-ip

Implemented the extended description for the check in cyhy-ip so it is known for later on why we are implementing it this way

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Re-worked the logic in geoloc.py so it now searches each IP address in the cidr, instead of just the first one. I am still having it just print the original cidr to avoid excessive printing to the console in the event of a large cidr, but that should still give enough information to go on. I also added comments to that effect in the check_restricted_cidr function.

dav3r commented on cyhy/core/geoloc.py:
```diff
@@ -43,11 +43,16 @@ def lookup(self, ip):
+            try:
+                response = self.__reader.city(str(ip))
+                if response.country.name in RESTRICTED_COUNTRIES:
+                    has_restricted = True
```
@st0rmbl3ss3d - One way to improve this would be to add each restricted IP address to an IPSet, then output the contents of that IPSet at the end.  The IPSet will automatically consolidate the list of addresses into its minimal CIDR form.

For reference:
* https://netaddr.readthedocs.io/en/latest/tutorial_03.html
* https://github.com/cisagov/cyhy-core/search?q=IPSet

st0rmbl3ss3d committed:
added an import of IPSet to use with the check_restricted_cidr function. Had each IP in the cidr that is found to be in one of the countries to the restricted_ips IPSet, and printed them out at the end. I chaged up the verbiage on the print statement to match what is happening. Also, since we are using the restricted_ips IPSet to detemrine if there are any restricted IPs, I removed the has_restricted variable and had the funciton return True if the length of restricted_ips is greater than zero. Otherwise, it will return False.

dav3r commented on cyhy/core/geoloc.py:
```diff
@@ -4,13 +4,15 @@
 GEODB_FILES = ["GeoIP2-City.mmdb", "GeoLite2-City.mmdb"]
 GEODB_CITY_PATHS = ["/usr/share/GeoIP/", "/usr/local/share/GeoIP/"]
 GEODB_FILE_PATHS = []
+RESTRICTED_COUNTRIES = ["China","Iran","North Korea","Russia"]
```
Spaces are preferred between list items.  Normally we would run this whole file thru `black` to format it, but that is not necessary for this PR.
```suggestion
RESTRICTED_COUNTRIES = ["China", "Iran", "North Korea", "Russia"]
```

dav3r commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +41,22 @@ def lookup(self, ip):
+                    restricted_ips.add(ip)
+            except AddressNotFoundError:
+                print >> sys.stderr, "CIDR %s not found in geolocation database" % cidr
+        # to avoid printing each and every ip in a cidr that is found
+        # this is only printing the cidr in which one or more addresses are found
+        # that correspond to a restricted country
```
This comment is a bit misleading since we are now printing the list of CIDRs representing every restricted IP found, not "the CIDRs in which one or more addresses are [restricted]".  Can you please clean up the wording here?

st0rmbl3ss3d replied:
Ahh right, I changed the functionality without the comment. Thanks

st0rmbl3ss3d replied:
So, now that the printing works as expected, just printing all the networks that are found, do we need a comment at all here? Originally, the comment was there to show the possibly unexpected print results, but now putting a comment here explaining what is going on seems redundant. Thoughts?

dav3r replied:
Good point- I'm fine with removing that comment.

dav3r commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +41,22 @@ def lookup(self, ip):
+        # this is only printing the cidr in which one or more addresses are found
+        # that correspond to a restricted country
+        if len(restricted_ips) > 0:
+            print >> sys.stderr, "Warning! The follwing networks can be traced to restricted country: %s" % response.country.name
```
Typo:
```suggestion
            print >> sys.stderr, "Warning! The following networks can be traced to restricted country: %s" % response.country.name
```

st0rmbl3ss3d replied:
Oops! Good catch.

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

Adding spaces between list items in RESTRICTED_COUNTRIES to confirm with preferred format.

Co-authored-by: dav3r <daver@geekpad.com>

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

fixed typo in error message

Co-authored-by: dav3r <daver@geekpad.com>

st0rmbl3ss3d committed:
removing the comments here as they are no longer necessary since the execution now works as expected.

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -4,13 +4,15 @@
 
 import os
 import sys
+from netaddr import IPSet
```
```suggestion
```

st0rmbl3ss3d replied:
Ahh yeah this is a more elegant implementation, I think. Good idea. I will need to go into the checks in ip-check and import to re-work how they use it then.

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +41,19 @@ def lookup(self, ip):
         except AddressNotFoundError:
             return (None, None)
+
+    def check_restricted_cidr(self, cidr):
+        restricted_ips = IPSet()
+        for ip in cidr:
+            try:
+                response = self.__reader.city(str(ip))
+                if response.country.name in RESTRICTED_COUNTRIES:
+                    restricted_ips.add(ip)
+            except AddressNotFoundError:
+                print >> sys.stderr, "CIDR %s not found in geolocation database" % cidr
+        if len(restricted_ips) > 0:
+            print >> sys.stderr, "Warning! The following networks can be traced to restricted country: %s" % response.country.name
+            for network in restricted_ips.iter_cidrs():
+                print >> sys.stderr, network
+            return True
+        return False
```
```suggestion
    def check_restricted_ip(self, ip):
        try:
            response = self.__reader.city(str(ip))
            if response.country.name in RESTRICTED_COUNTRIES:
                return True
            return False
        except AddressNotFoundError:
            print >> sys.stderr, "CIDR %s not found in geolocation database" % cidr
```

st0rmbl3ss3d replied:
I will commit this suggestion, but will slightly alter it in a different commit. The wording should be changed to indicate the IP is not found in the geoloc db instead of the cidr.

st0rmbl3ss3d replied:
Actually, thinking about it more... one thing I did like about the current implementation is the ability to point out which country it was traced to. If that is not necessary, then I won't worry about it, but if we do want that functionality, I could have it return the country as well as the true/false possibly, but that might be a little clunky. Bottom line though, is do we care about knowing which country it is from or do we only care that it is in the list?

mcdonnnj replied:
I'd be ok with returning a tuple of

```python
(boolean, "Country name or empty string")
```

if you wanted that information. You could do a dict of `"Country Name": IPSet()` key/values in the scripts to minimize output.

mcdonnnj replied:
Additional thought I had is you could consider using [`namedtuple()`](https://docs.python.org/2.7/library/collections.html#collections.namedtuple) to manage the return value. I use these when I want to provide a return that contains multiple values. An example is in [cisagov/hash-http-content](https://github.com/cisagov/hash-http-content/blob/4050828fd6b74a080d156cd668fceec0e4e9414f/src/hash_http_content/hasher.py#L64-L79), but note the syntax is different because that is a Python 3 project.

mcdonnnj reviewed (changes requested):
Sorry for further trouble, but I bring more changes to request. I *strongly* prefer reducing the complexity in the method added to `GeoLocDB`. I think that it should take a single IP and check it, returning a boolean value and only printing output if an exception occurs. I would like the scripts to do the heavy lifting for iterating over input IPs, building an IPSet of restricted IPs, and handling output to inform the user. I apologize again as this is more work, but I think that this approach aligns more with the design of this library, and more specifically the `GeoLocDB` class.

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

removing the import to IPSet because we are adjusting the implementation of the check_restricted_ip function to instead check a single IP and return true/false. Then we can implement the iteration and adding to an IPSet on the side of where the check is being called

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

Reducing complexity of this function by having it only check a single IP and moving the looping through a CIDR to the calling of the function in cyhy-ip and cyhy-import. Will edit later to have it return Country names with the IPs

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Changed what is returned in geoloc.py to True and response.country.name if the country is in the restricted list and to return False and None if not. Adjusted the workings of the function in cyhy-ip to handle the multiple returns and it now creates a defaultdict to handle multiple countries if present and will print out a list of countries and their corresponding IPs. Removed the has_restriced variable as now the length of the dict being > 0 can be used instead as a a determinant. Added an import for defaultdict so the functionality can be used.

st0rmbl3ss3d commented:
Just committed some changes. What do you think of that handling? Or should we use something other than defaultdict?

jsf9k reviewed (approved):
Now approved even harder.

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +40,12 @@ def lookup(self, ip):
+            response = self.__reader.city(str(ip))
+            if response.country.name in RESTRICTED_COUNTRIES:
+                return True,response.country.name
+            return False,None
```
The `False` return has been moved to the end of the method.

```suggestion
                return True, response.country.name
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +163,25 @@ def add(db, owner, cidrs):
+    geo_loc_db = GeoLocDB()
+    restricted_dict = defaultdict(lambda:IPSet())
+    for ip in cidrs:
+        is_restricted,country = geo_loc_db.check_restricted_ip(ip)
```
```suggestion
        is_restricted, country = geo_loc_db.check_restricted_ip(ip)
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +163,25 @@ def add(db, owner, cidrs):
+        for key,value in restricted_dict.items():
+            print("%s:" % key)
+            for cidr in value.iter_cidrs():
+                print("    %s" % cidr)
```
Making the `print` statements consistent with the rest of this project, fixing some formatting, and using more helpful variable names.

```suggestion
        for country, cidrs in restricted_dict.items():
            print "%s:" % country
            for cidr in cidrs.iter_cidrs():
                print "    %s" % cidr
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +163,25 @@ def add(db, owner, cidrs):
         cidrs
     )  # intersections from database
+    geo_loc_db = GeoLocDB()
+    restricted_dict = defaultdict(lambda:IPSet())
```
```suggestion
    restricted_dict = defaultdict(lambda: IPSet())
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +163,25 @@ def add(db, owner, cidrs):
+            # want the user to see all restricted CIDR blocks
+            # without having to run multiple times.
+    if len(restricted_dict) > 0:
+        print("***Found IPs in restricted countries***")
```
Consistent with other `print` statements in this project.

```suggestion
        print "***Found IPs in restricted countries***"
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +163,25 @@ def add(db, owner, cidrs):
+            print("%s:" % key)
+            for cidr in value.iter_cidrs():
+                print("    %s" % cidr)
+        print "Cannot continue!\nSome addresses traced to restricted countries."
```
I feel liked "traced" implies a more thorough check than we're performing.

```suggestion
        print "Cannot continue!\nSome addresses associated with restricted countries."
```

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +40,12 @@ def lookup(self, ip):
+                return True,response.country.name
+            return False,None
+        except AddressNotFoundError:
+            print >> sys.stderr, "CIDR %s not found in geolocation database" % cidr
```
So that this method has a return regardless of an exception. The return type for this method should be a `Tuple[bool, str]`, so I replaced the `None` with an empty string while moving the return statement.

```suggestion
            print >> sys.stderr, "CIDR %s not found in geolocation database" % cidr

        return False, ""
```

mcdonnnj reviewed (changes requested):
Thanks for the continued work on polishing this to completion. I had some changes that are *mostly* just formatting. One is a wording suggestion. Lastly there's some adjustments to the new method in `geoloc.py`.

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

implemented the suggestion to move the False statement in geoloc.py and reformat

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

implemented suggestions for formatting how the variables are assigned in cyhy-ip by adding a space between variables

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

implemented suggested changes. The variables names are more explicit, which is good. Also the formatting is fixed to match the rest of the code. I do a lot of other things in python3 so I keep defaulting to the python3 printing format haha

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

added the space here too for the formatting consistency.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

updated the formatting on this print statement to match the others

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

Agreed. The wording here is better for what we are doing. Committing.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update cyhy/core/geoloc.py

Sounds good. An empty string should be fine as well, but will do some testing to make sure.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Forgot to change cyhy-import to match the new format. Updated the refrences and changed the logic to mirror what cyhy-ip is doing with the defaultdict and the checking of IPs. removed the old checking methods.

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -26,13 +26,15 @@ import sys
 from cyhy.core import Config, STATUS, STAGE
 from cyhy.db import database
 from cyhy.util import util
+from cyhy.core.geoloc import GeoLocDB
```
Fixing import ordering.

```suggestion
from cyhy.core.geoloc import GeoLocDB
from cyhy.db import database
from cyhy.util import util
```

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -26,13 +26,15 @@ import sys
 # Third-Party Libraries
 import dateutil
 from docopt import docopt
+from collections import defaultdict
```
This is a standard library for Python. This should be up in the appropriate section.

st0rmbl3ss3d replied:
Roger. Will move to the correct section

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -79,6 +81,27 @@ def has_intersections(db, nets, filename, owner):
+            # prints output for each restricted CIDR block and we
+            # want the user to see all restricted CIDR blocks
+            # without having to run multiple times.
+    if len(restricted_dict) > 0:
```
This should have correct truthy-ness if it is empty or has elements.

```suggestion
    if restricted_dict:
```

```console
>>> test_dict = defaultdict(lambda: netaddr.IPSet())
>>> print True if test_dict else False
False
>>> test_dict["Test"].add("192.168.1.1")
>>> print True if test_dict else False
True
```

jsf9k replied:
I personally prefer the more specific code that is already present, although @mcdonnnj's suggestion is more Pythonic.  I'm OK with either way.  Is this preference addressed (possibly via a link to another document) in our development guide docs?

mcdonnnj replied:
The development guide doesn't say much about actual Python code preferences. When dealing with things like this I prefer to defer to the language community's customs so that you can leverage optimizations and have code that appears similar to other code written in the same language.

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -26,6 +26,7 @@ Notes:
 import sys
 import progressbar as pb
 from docopt import docopt
+from collections import defaultdict
```
These haven't been broke up correctly, but I would at least like this import grouped with the other standard library import.

```suggestion
from collections import defaultdict
import sys
import progressbar as pb
from docopt import docopt
```

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -163,6 +163,25 @@ def add(db, owner, cidrs):
+            # prints output for each restricted CIDR block and we
+            # want the user to see all restricted CIDR blocks
+            # without having to run multiple times.
+    if len(restricted_dict) > 0:
```
Same comment about truthy-ness.

```suggestion
    if restricted_dict:
```

mcdonnnj commented on cyhy/core/geoloc.py:
```diff
@@ -39,3 +40,13 @@ def lookup(self, ip):
         except AddressNotFoundError:
             return (None, None)
+
+    def check_restricted_ip(self, ip):
```
With other changes in how this method is used, is there a compelling reason to not just return a string? This would simplify this method and how it is used because it would be an empty string or a restricted country. Example showing the truthy-ness:

```console
>>> test_string = ""
>>> print True if test_string else False
False
>>> test_string = "Hello, World!"
>>> print True if test_string else False
True
```

jsf9k replied:
This makes the code more Pythonic, at (IMHO) the expense of clarity.  I'm fine either way.

st0rmbl3ss3d replied:
It would simplify things having just one return instead of two. I didn't realize the empty string/dict would come up as false, so I think we can change it that way.

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -26,13 +26,15 @@ import sys
 from docopt import docopt
+from collections import defaultdict
 import progressbar as pb
 import netaddr
```
I would like this to be updated correspondingly as this is the only thing we're using. Uses of `netaddr.IPSet` would need to be updated as well.

```suggestion
from netaddr import IPSet
```

mcdonnnj reviewed (changes requested):
Some import and truthy-ness changes. I also had a question about the `GeoLocDB.check_restricted_ip()` method's return value based on other changes that have been made in this PR.

st0rmbl3ss3d committed:
Update bin/cyhy-import

committing changes to import ordering to be more consistent

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

committing suggestion for more consistency. grouping standard libraries together

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-import

changing the import to only IPSet as it is the only function being used

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-import

shortening the logic for checking if there are things in the restricted dict.

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

updating to match cyhy-import with the shortening of the check logic

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Moved the import of defaultdict in cyhy-import to the proper section. Reworked check_restricted_ip in geoloc.py to only return one element. It will return a blank string if there is no country name. Also updated the logic in cyhy-ip and cyhy-import to accept the single return as well.

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -22,15 +22,17 @@ Notes:
 import datetime
 import json
 import sys
+from collections import defaultdict
```
```suggestion
from collections import defaultdict
import datetime
import json
import sys
```

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -22,15 +22,17 @@ Notes:
 from docopt import docopt
 import progressbar as pb
-import netaddr
+from netaddr import IPSet
```
This should be above the `progressbar` import.

st0rmbl3ss3d replied:
implementing import of IPSet over all of netaddr

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -6,23 +6,24 @@ Usage:
 
+from collections import defaultdict
 import sys
 import progressbar as pb
 from docopt import docopt
```
```suggestion
from docopt import docopt
import progressbar as pb
```

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -79,6 +81,27 @@ def has_intersections(db, nets, filename, owner):
 
+def has_restricted_ips(nets):
+    geo_loc_db = GeoLocDB()
+    restricted_dict = defaultdict(lambda: netaddr.IPSet())
```
The other use of `netaddr.IPSet()` at https://github.com/cisagov/cyhy-core/blob/127fdccc1c21181f2fb60a145a7bd75ff1199ad4/bin/cyhy-import#L114 needs to be done as well. This is needed because of the change to the import.

```suggestion
    restricted_dict = defaultdict(lambda: IPSet())
```

mcdonnnj commented on bin/cyhy-import:
```diff
@@ -79,6 +81,27 @@ def has_intersections(db, nets, filename, owner):
         print_intersections(intersections)
         return True
 
+def has_restricted_ips(nets):
```
```suggestion

def has_restricted_ips(nets):
```

mcdonnnj reviewed (changes requested):
Just some import changes and a small formatting change.

st0rmbl3ss3d committed:
Update bin/cyhy-import

implementing import change for proper formatting

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-ip

Not sure if I changed anything or if this is in there form before, but implementing this for consistency

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-import

changing call to IPSet with import change

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

st0rmbl3ss3d committed:
Update bin/cyhy-import

implementing change to add blank lines for formatting

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

Commit:
changing import order for consistency

mcdonnnj commented on bin/cyhy-ip:
```diff
@@ -6,26 +6,27 @@ Usage:
   and the --file option is omitted.
 """
 
+from collections import defaultdict
```
The last thing is if you would please organize the imports in this file to mirror that in `bin/cyhy-import`. It would be very, very appreciated.

jsf9k replied:
@mcdonnnj = Detective Columbo :smile: 
![image](https://user-images.githubusercontent.com/5521725/121550951-363f3700-c9dd-11eb-9e2d-d345ed7d9648.png)

st0rmbl3ss3d replied:
Just pushed an update with the import changes, @mcdonnnj. Sorry it took a while, it's been a hectic week. Thanks!

mcdonnnj reviewed (changes requested):
One last change that would be greatly appreciated. Thanks again for all the tolerance of my change request. As soon as that is done please reach out to me so I can slam the approve button for you.

Commit:
Updating the imports to mirror that of bin/cyhy-import. This should add more consistency across utilities and how they are organized.

dav3r reviewed (approved):
This all looks good to my eyes! 👀 
Nice work @st0rmbl3ss3d!

jsf9k reviewed (approved):
Great work!  Amazing stamina and patience, @st0rmbl3ss3d, for sticking with us this far.

mcdonnnj reviewed (approved):
Let's get this weary traveler home!

mcdonnnj commented:
@st0rmbl3ss3d Sorry this slid by this week, but you should be good to build a new image now.
